File: src/vault/DonkeyWork.Vault.Contracts/CredentialKind.cs
using System.Text.Json.Serialization;

namespace DonkeyWork.Vault.Contracts;

/// <summary>
/// Explicit credential kind — the discriminator that tells an agent how to use the stored
/// secret. Serialized as a snake_case string on both the API (JSON) and the entity (a text
/// column via <see cref="CredentialKindExtensions"/>), never as an integer.
/// </summary>
[JsonConverter(typeof(JsonStringEnumConverter<CredentialKind>))]
public enum CredentialKind
{
    /// <summary>Catch-all: the secret is opaque material, returned verbatim (HMAC secrets, tokens with no header, …).</summary>
    [JsonStringEnumMemberName("opaque")]
    Opaque = 0,

    /// <summary>API key sent in an HTTP header: <c>{header}: {prefix}{secret}</c>.</summary>
    [JsonStringEnumMemberName("header_api_key")]
    HeaderApiKey,

    /// <summary>HTTP Basic: <c>Authorization: Basic base64(username:secret)</c>.</summary>
    [JsonStringEnumMemberName("http_basic")]
    HttpBasic,

    /// <summary>A username + password login that is NOT sent as HTTP Basic (e.g. an OAuth
    /// ROPC password grant, a DSM/query-param login, a DB user) — username is metadata,
    /// the secret is the password; no header is assembled.</summary>
    [JsonStringEnumMemberName("username_password")]
    UsernamePassword,

    /// <summary>SSH login: username + host (base_url = <c>ssh://host:port</c>); secret is the password or key.</summary>
    [JsonStringEnumMemberName("ssh")]
    Ssh,

    /// <summary>The whole connection string / DSN is the secret; returned verbatim.</summary>
    [JsonStringEnumMemberName("connection_string")]
    ConnectionString,
}

/// <summary>Snake_case wire mapping for <see cref="CredentialKind"/>, used for the entity's text column.</summary>
public static class CredentialKindExtensions
{
    public static string ToWire(this CredentialKind kind) => kind switch
    {
        CredentialKind.HeaderApiKey => "header_api_key",
        CredentialKind.HttpBasic => "http_basic",
        CredentialKind.UsernamePassword => "username_password",
        CredentialKind.Ssh => "ssh",
        CredentialKind.ConnectionString => "connection_string",
        _ => "opaque",
    };

    public static CredentialKind FromWire(string? wire) => wire switch
    {
        "header_api_key" => CredentialKind.HeaderApiKey,
        "http_basic" => CredentialKind.HttpBasic,
        "username_password" => CredentialKind.UsernamePassword,
        "ssh" => CredentialKind.Ssh,
        "connection_string" => CredentialKind.ConnectionString,
        _ => CredentialKind.Opaque,
    };
}

File: src/server/internal/contracts/credentialkind.go
package contracts

import (
	"encoding/json"
	"fmt"
)

// CredentialKind is the discriminator that tells an agent how to use a stored secret. It is
// serialized as a snake_case string on both the API (JSON) and the entity (a text column), never as
// an integer — the wire form the CLI/SPA consume.
type CredentialKind string

const (
	// KindOpaque is the catch-all: the secret is opaque material, returned verbatim.
	KindOpaque CredentialKind = "opaque"
	// KindHeaderAPIKey is sent in an HTTP header: {header}: {prefix}{secret}.
	KindHeaderAPIKey CredentialKind = "header_api_key"
	// KindHTTPBasic is HTTP Basic: Authorization: Basic base64(username:secret).
	KindHTTPBasic CredentialKind = "http_basic"
	// KindUsernamePassword is a username+password login that is NOT sent as HTTP Basic.
	KindUsernamePassword CredentialKind = "username_password"
	// KindSSH is an SSH login: username + host; secret is the password or key.
	KindSSH CredentialKind = "ssh"
	// KindConnectionString means the whole connection string / DSN is the secret.
	KindConnectionString CredentialKind = "connection_string"
)

// CredentialKindFromWire maps a stored/text value to a CredentialKind, defaulting to opaque for any
// unknown or empty input.
func CredentialKindFromWire(s string) CredentialKind {
	switch CredentialKind(s) {
	case KindHeaderAPIKey, KindHTTPBasic, KindUsernamePassword, KindSSH, KindConnectionString:
		return CredentialKind(s)
	default:
		return KindOpaque
	}
}

// MarshalJSON always emits a known wire value (defaulting unknowns to opaque).
func (k CredentialKind) MarshalJSON() ([]byte, error) {
	return json.Marshal(string(CredentialKindFromWire(string(k))))
}

// UnmarshalJSON accepts the snake_case wire form, defaulting unknowns to opaque.
func (k *CredentialKind) UnmarshalJSON(b []byte) error {
	var s string
	if err := json.Unmarshal(b, &s); err != nil {
		return fmt.Errorf("credential kind must be a string: %w", err)
	}
	*k = CredentialKindFromWire(s)
	return nil
}
